### PR TITLE
Updated error handling for templating library

### LINKED ISSUE #3976: [.NET][Templating] MSApprovals sample card triggers NullReferenceException in AdaptiveCards.Templating.dll
# Platform

* .NET Templating SDK

# Version of SDK

Outbound Preview under-test

# Details
[Current_sampleData.zip](https://github.com/microsoft/AdaptiveCards/files/4561157/Current_sampleData.zip)

The attached sample card template as well as the card data trigger this exception in our library: I believe we need to check GetCurrentDataContext() for null before trying to access any properties on it.

```
System.NullReferenceException
  HResult=0x80004003
  Message=Object reference not set to an instance of an object.
  Source=AdaptiveCards.Templating
  StackTrace:
   at AdaptiveCards.Templating.AdaptiveCardsTemplateVisitor.VisitObj(ObjContext context) in C:\AdaptiveCards\source\dotnet\Library\AdaptiveCards.Templating\AdaptiveCardsTemplateVisitor.cs:line 381
```

Callstack:
```
>	AdaptiveCards.Templating.dll!AdaptiveCards.Templating.AdaptiveCardsTemplateVisitor.VisitObj(AdaptiveCards.Templating.AdaptiveCardsTemplateParser.ObjContext context) Line 381	C#
 	AdaptiveCards.Templating.dll!AdaptiveCards.Templating.AdaptiveCardsTemplateVisitor.VisitValueObject(AdaptiveCards.Templating.AdaptiveCardsTemplateParser.ValueObjectContext context) Line 349	C#
 	AdaptiveCards.Templating.dll!AdaptiveCards.Templating.AdaptiveCardsTemplateParser.ValueObjectContext.Accept<AdaptiveCards.Templating.AdaptiveCardsTemplateResult>(Antlr4.Runtime.Tree.IParseTreeVisitor<AdaptiveCards.Templating.AdaptiveCardsTemplateResult> visitor) Line 531	C#
 	Antlr4.Runtime.dll!Antlr4.Runtime.Tree.AbstractParseTreeVisitor<AdaptiveCards.Templating.AdaptiveCardsTemplateResult>.Visit(Antlr4.Runtime.Tree.IParseTree tree)	Unknown
 	AdaptiveCards.Templating.dll!AdaptiveCards.Templating.AdaptiveCardsTemplateVisitor.VisitArray(AdaptiveCards.Templating.AdaptiveCardsTemplateParser.ArrayContext context) Line 617	C#
 	AdaptiveCards.Templating.dll!AdaptiveCards.Templating.AdaptiveCardsTemplateParser.ArrayContext.Accept<AdaptiveCards.Templating.AdaptiveCardsTemplateResult>(Antlr4.Runtime.Tree.IParseTreeVisitor<AdaptiveCards.Templating.AdaptiveCardsTemplateResult> visitor) Line 403	C#
 	Antlr4.Runtime.dll!Antlr4.Runtime.Tree.AbstractParseTreeVisitor<AdaptiveCards.Templating.AdaptiveCardsTemplateResult>.Visit(Antlr4.Runtime.Tree.IParseTree tree)	Unknown
 	AdaptiveCards.Templating.dll!AdaptiveCards.Templating.AdaptiveCardsTemplateVisitor.VisitChildren(Antlr4.Runtime.Tree.IRuleNode node) Line 663	C#
 	AdaptiveCards.Templating.dll!AdaptiveCards.Templating.AdaptiveCardsTemplateParserBaseVisitor<AdaptiveCards.Templating.AdaptiveCardsTemplateResult>.VisitValueArray(AdaptiveCards.Templating.AdaptiveCardsTemplateParser.ValueArrayContext context) Line 155	C#
 	AdaptiveCards.Templating.dll!AdaptiveCards.Templating.AdaptiveCardsTemplateParser.ValueArrayContext.Accept<AdaptiveCards.Templating.AdaptiveCardsTemplateResult>(Antlr4.Runtime.Tree.IParseTreeVisitor<AdaptiveCards.Templating.AdaptiveCardsTemplateResult> visitor) Line 542	C#
 	Antlr4.Runtime.dll!Antlr4.Runtime.Tree.AbstractParseTreeVisitor<AdaptiveCards.Templating.AdaptiveCardsTemplateResult>.Visit(Antlr4.Runtime.Tree.IParseTree tree)	Unknown
 	AdaptiveCards.Templating.dll!AdaptiveCards.Templating.AdaptiveCardsTemplateVisitor.VisitChildren(Antlr4.Runtime.Tree.IRuleNode node) Line 663	C#
 	AdaptiveCards.Templating.dll!AdaptiveCards.Templating.AdaptiveCardsTemplateParserBaseVisitor<AdaptiveCards.Templating.AdaptiveCardsTemplateResult>.VisitJsonPair(AdaptiveCards.Templating.AdaptiveCardsTemplateParser.JsonPairContext context) Line 47	C#
 	AdaptiveCards.Templating.dll!AdaptiveCards.Templating.AdaptiveCardsTemplateParser.JsonPairContext.Accept<AdaptiveCards.Templating.AdaptiveCardsTemplateResult>(Antlr4.Runtime.Tree.IParseTreeVisitor<AdaptiveCards.Templating.AdaptiveCardsTemplateResult> visitor) Line 244	C#
 	Antlr4.Runtime.dll!Antlr4.Runtime.Tree.AbstractParseTreeVisitor<AdaptiveCards.Templating.AdaptiveCardsTemplateResult>.Visit(Antlr4.Runtime.Tree.IParseTree tree)	Unknown
 	AdaptiveCards.Templating.dll!AdaptiveCards.Templating.AdaptiveCardsTemplateVisitor.VisitObj(AdaptiveCards.Templating.AdaptiveCardsTemplateParser.ObjContext context) Line 422	C#
 	AdaptiveCards.Templating.dll!AdaptiveCards.Templating.AdaptiveCardsTemplateVisitor.VisitValueObject(AdaptiveCards.Templating.AdaptiveCardsTemplateParser.ValueObjectContext context) Line 349	C#
 	AdaptiveCards.Templating.dll!AdaptiveCards.Templating.AdaptiveCardsTemplateParser.ValueObjectContext.Accept<AdaptiveCards.Templating.AdaptiveCardsTemplateResult>(Antlr4.Runtime.Tree.IParseTreeVisitor<AdaptiveCards.Templating.AdaptiveCardsTemplateResult> visitor) Line 531	C#
 	Antlr4.Runtime.dll!Antlr4.Runtime.Tree.AbstractParseTreeVisitor<AdaptiveCards.Templating.AdaptiveCardsTemplateResult>.Visit(Antlr4.Runtime.Tree.IParseTree tree)	Unknown
 	AdaptiveCards.Templating.dll!AdaptiveCards.Templating.AdaptiveCardsTemplateVisitor.VisitArray(AdaptiveCards.Templating.AdaptiveCardsTemplateParser.ArrayContext context) Line 617	C#
 	AdaptiveCards.Templating.dll!AdaptiveCards.Templating.AdaptiveCardsTemplateParser.ArrayContext.Accept<AdaptiveCards.Templating.AdaptiveCardsTemplateResult>(Antlr4.Runtime.Tree.IParseTreeVisitor<AdaptiveCards.Templating.AdaptiveCardsTemplateResult> visitor) Line 403	C#
 	Antlr4.Runtime.dll!Antlr4.Runtime.Tree.AbstractParseTreeVisitor<AdaptiveCards.Templating.AdaptiveCardsTemplateResult>.Visit(Antlr4.Runtime.Tree.IParseTree tree)	Unknown
 	AdaptiveCards.Templating.dll!AdaptiveCards.Templating.AdaptiveCardsTemplateVisitor.VisitChildren(Antlr4.Runtime.Tree.IRuleNode node) Line 663	C#
 	AdaptiveCards.Templating.dll!AdaptiveCards.Templating.AdaptiveCardsTemplateParserBaseVisitor<AdaptiveCards.Templating.AdaptiveCardsTemplateResult>.VisitValueArray(AdaptiveCards.Templating.AdaptiveCardsTemplateParser.ValueArrayContext context) Line 155	C#
 	AdaptiveCards.Templating.dll!AdaptiveCards.Templating.AdaptiveCardsTemplateParser.ValueArrayContext.Accept<AdaptiveCards.Templating.AdaptiveCardsTemplateResult>(Antlr4.Runtime.Tree.IParseTreeVisitor<AdaptiveCards.Templating.AdaptiveCardsTemplateResult> visitor) Line 542	C#
 	Antlr4.Runtime.dll!Antlr4.Runtime.Tree.AbstractParseTreeVisitor<AdaptiveCards.Templating.AdaptiveCardsTemplateResult>.Visit(Antlr4.Runtime.Tree.IParseTree tree)	Unknown
 	AdaptiveCards.Templating.dll!AdaptiveCards.Templating.AdaptiveCardsTemplateVisitor.VisitChildren(Antlr4.Runtime.Tree.IRuleNode node) Line 663	C#
 	AdaptiveCards.Templating.dll!AdaptiveCards.Templating.AdaptiveCardsTemplateParserBaseVisitor<AdaptiveCards.Templating.AdaptiveCardsTemplateResult>.VisitJsonPair(AdaptiveCards.Templating.AdaptiveCardsTemplateParser.JsonPairContext context) Line 47	C#
 	AdaptiveCards.Templating.dll!AdaptiveCards.Templating.AdaptiveCardsTemplateParser.JsonPairContext.Accept<AdaptiveCards.Templating.AdaptiveCardsTemplateResult>(Antlr4.Runtime.Tree.IParseTreeVisitor<AdaptiveCards.Templating.AdaptiveCardsTemplateResult> visitor) Line 244	C#
 	Antlr4.Runtime.dll!Antlr4.Runtime.Tree.AbstractParseTreeVisitor<AdaptiveCards.Templating.AdaptiveCardsTemplateResult>.Visit(Antlr4.Runtime.Tree.IParseTree tree)	Unknown
 	AdaptiveCards.Templating.dll!AdaptiveCards.Templating.AdaptiveCardsTemplateVisitor.VisitObj(AdaptiveCards.Templating.AdaptiveCardsTemplateParser.ObjContext context) Line 422	C#
 	AdaptiveCards.Templating.dll!AdaptiveCards.Templating.AdaptiveCardsTemplateVisitor.VisitValueObject(AdaptiveCards.Templating.AdaptiveCardsTemplateParser.ValueObjectContext context) Line 349	C#
 	AdaptiveCards.Templating.dll!AdaptiveCards.Templating.AdaptiveCardsTemplateParser.ValueObjectContext.Accept<AdaptiveCards.Templating.AdaptiveCardsTemplateResult>(Antlr4.Runtime.Tree.IParseTreeVisitor<AdaptiveCards.Templating.AdaptiveCardsTemplateResult> visitor) Line 531	C#
 	Antlr4.Runtime.dll!Antlr4.Runtime.Tree.AbstractParseTreeVisitor<AdaptiveCards.Templating.AdaptiveCardsTemplateResult>.Visit(Antlr4.Runtime.Tree.IParseTree tree)	Unknown
 	AdaptiveCards.Templating.dll!AdaptiveCards.Templating.AdaptiveCardsTemplateVisitor.VisitArray(AdaptiveCards.Templating.AdaptiveCardsTemplateParser.ArrayContext context) Line 617	C#
 	AdaptiveCards.Templating.dll!AdaptiveCards.Templating.AdaptiveCardsTemplateParser.ArrayContext.Accept<AdaptiveCards.Templating.AdaptiveCardsTemplateResult>(Antlr4.Runtime.Tree.IParseTreeVisitor<AdaptiveCards.Templating.AdaptiveCardsTemplateResult> visitor) Line 403	C#
 	Antlr4.Runtime.dll!Antlr4.Runtime.Tree.AbstractParseTreeVisitor<AdaptiveCards.Templating.AdaptiveCardsTemplateResult>.Visit(Antlr4.Runtime.Tree.IParseTree tree)	Unknown
 	AdaptiveCards.Templating.dll!AdaptiveCards.Templating.AdaptiveCardsTemplateVisitor.VisitChildren(Antlr4.Runtime.Tree.IRuleNode node) Line 663	C#
 	AdaptiveCards.Templating.dll!AdaptiveCards.Templating.AdaptiveCardsTemplateParserBaseVisitor<AdaptiveCards.Templating.AdaptiveCardsTemplateResult>.VisitValueArray(AdaptiveCards.Templating.AdaptiveCardsTemplateParser.ValueArrayContext context) Line 155	C#
 	AdaptiveCards.Templating.dll!AdaptiveCards.Templating.AdaptiveCardsTemplateParser.ValueArrayContext.Accept<AdaptiveCards.Templating.AdaptiveCardsTemplateResult>(Antlr4.Runtime.Tree.IParseTreeVisitor<AdaptiveCards.Templating.AdaptiveCardsTemplateResult> visitor) Line 542	C#
 	Antlr4.Runtime.dll!Antlr4.Runtime.Tree.AbstractParseTreeVisitor<AdaptiveCards.Templating.AdaptiveCardsTemplateResult>.Visit(Antlr4.Runtime.Tree.IParseTree tree)	Unknown
 	AdaptiveCards.Templating.dll!AdaptiveCards.Templating.AdaptiveCardsTemplateVisitor.VisitChildren(Antlr4.Runtime.Tree.IRuleNode node) Line 663	C#
 	AdaptiveCards.Templating.dll!AdaptiveCards.Templating.AdaptiveCardsTemplateParserBaseVisitor<AdaptiveCards.Templating.AdaptiveCardsTemplateResult>.VisitJsonPair(AdaptiveCards.Templating.AdaptiveCardsTemplateParser.JsonPairContext context) Line 47	C#
 	AdaptiveCards.Templating.dll!AdaptiveCards.Templating.AdaptiveCardsTemplateParser.JsonPairContext.Accept<AdaptiveCards.Templating.AdaptiveCardsTemplateResult>(Antlr4.Runtime.Tree.IParseTreeVisitor<AdaptiveCards.Templating.AdaptiveCardsTemplateResult> visitor) Line 244	C#
 	Antlr4.Runtime.dll!Antlr4.Runtime.Tree.AbstractParseTreeVisitor<AdaptiveCards.Templating.AdaptiveCardsTemplateResult>.Visit(Antlr4.Runtime.Tree.IParseTree tree)	Unknown
 	AdaptiveCards.Templating.dll!AdaptiveCards.Templating.AdaptiveCardsTemplateVisitor.VisitObj(AdaptiveCards.Templating.AdaptiveCardsTemplateParser.ObjContext context) Line 422	C#
 	AdaptiveCards.Templating.dll!AdaptiveCards.Templating.AdaptiveCardsTemplateVisitor.VisitValueObject(AdaptiveCards.Templating.AdaptiveCardsTemplateParser.ValueObjectContext context) Line 349	C#
 	AdaptiveCards.Templating.dll!AdaptiveCards.Templating.AdaptiveCardsTemplateParser.ValueObjectContext.Accept<AdaptiveCards.Templating.AdaptiveCardsTemplateResult>(Antlr4.Runtime.Tree.IParseTreeVisitor<AdaptiveCards.Templating.AdaptiveCardsTemplateResult> visitor) Line 531	C#
 	Antlr4.Runtime.dll!Antlr4.Runtime.Tree.AbstractParseTreeVisitor<AdaptiveCards.Templating.AdaptiveCardsTemplateResult>.Visit(Antlr4.Runtime.Tree.IParseTree tree)	Unknown
 	AdaptiveCards.Templating.dll!AdaptiveCards.Templating.AdaptiveCardsTemplateVisitor.VisitChildren(Antlr4.Runtime.Tree.IRuleNode node) Line 663	C#
 	AdaptiveCards.Templating.dll!AdaptiveCards.Templating.AdaptiveCardsTemplateParserBaseVisitor<AdaptiveCards.Templating.AdaptiveCardsTemplateResult>.VisitJson(AdaptiveCards.Templating.AdaptiveCardsTemplateParser.JsonContext context) Line 238	C#
 	AdaptiveCards.Templating.dll!AdaptiveCards.Templating.AdaptiveCardsTemplateParser.JsonContext.Accept<AdaptiveCards.Templating.AdaptiveCardsTemplateResult>(Antlr4.Runtime.Tree.IParseTreeVisitor<AdaptiveCards.Templating.AdaptiveCardsTemplateResult> visitor) Line 115	C#
 	Antlr4.Runtime.dll!Antlr4.Runtime.Tree.AbstractParseTreeVisitor<AdaptiveCards.Templating.AdaptiveCardsTemplateResult>.Visit(Antlr4.Runtime.Tree.IParseTree tree)	Unknown
 	AdaptiveCards.Templating.dll!AdaptiveCards.Templating.AdaptiveCardsTemplate.Expand(AdaptiveCards.Templating.AdaptiveCardsEvaluationContext context, System.Func<string, object> nullSubstitutionOption) Line 112	C#
 	WpfVisualizer.exe!WpfVisualizer.MainWindow.RenderCard() Line 148	C#
 	WpfVisualizer.exe!WpfVisualizer.MainWindow.Timer_Tick(object sender, System.EventArgs e) Line 111	C#
 	WindowsBase.dll!System.Windows.Threading.DispatcherTimer.FireTick(object unused)	Unknown
 	WindowsBase.dll!System.Windows.Threading.ExceptionWrapper.InternalRealCall(System.Delegate callback, object args, int numArgs)	Unknown
 	WindowsBase.dll!System.Windows.Threading.ExceptionWrapper.TryCatchWhen(object source, System.Delegate callback, object args, int numArgs, System.Delegate catchHandler)	Unknown
 	WindowsBase.dll!System.Windows.Threading.DispatcherOperation.InvokeImpl()	Unknown
 	WindowsBase.dll!System.Windows.Threading.DispatcherOperation.InvokeInSecurityContext(object state)	Unknown
 	WindowsBase.dll!MS.Internal.CulturePreservingExecutionContext.CallbackWrapper(object obj)	Unknown
 	mscorlib.dll!System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state, bool preserveSyncCtx) Line 980	C#
 	mscorlib.dll!System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state, bool preserveSyncCtx) Line 928	C#
 	mscorlib.dll!System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state) Line 917	C#
 	WindowsBase.dll!MS.Internal.CulturePreservingExecutionContext.Run(MS.Internal.CulturePreservingExecutionContext executionContext, System.Threading.ContextCallback callback, object state)	Unknown
 	WindowsBase.dll!System.Windows.Threading.DispatcherOperation.Invoke()	Unknown
 	WindowsBase.dll!System.Windows.Threading.Dispatcher.ProcessQueue()	Unknown
 	WindowsBase.dll!System.Windows.Threading.Dispatcher.WndProcHook(System.IntPtr hwnd, int msg, System.IntPtr wParam, System.IntPtr lParam, ref bool handled)	Unknown
 	WindowsBase.dll!MS.Win32.HwndWrapper.WndProc(System.IntPtr hwnd, int msg, System.IntPtr wParam, System.IntPtr lParam, ref bool handled)	Unknown
 	WindowsBase.dll!MS.Win32.HwndSubclass.DispatcherCallbackOperation(object o)	Unknown
 	WindowsBase.dll!System.Windows.Threading.ExceptionWrapper.InternalRealCall(System.Delegate callback, object args, int numArgs)	Unknown
 	WindowsBase.dll!System.Windows.Threading.ExceptionWrapper.TryCatchWhen(object source, System.Delegate callback, object args, int numArgs, System.Delegate catchHandler)	Unknown
 	WindowsBase.dll!System.Windows.Threading.Dispatcher.LegacyInvokeImpl(System.Windows.Threading.DispatcherPriority priority, System.TimeSpan timeout, System.Delegate method, object args, int numArgs)	Unknown
 	WindowsBase.dll!MS.Win32.HwndSubclass.SubclassWndProc(System.IntPtr hwnd, int msg, System.IntPtr wParam, System.IntPtr lParam)	Unknown
 	[Native to Managed Transition]	
 	[Managed to Native Transition]	
 	WindowsBase.dll!System.Windows.Threading.Dispatcher.PushFrameImpl(System.Windows.Threading.DispatcherFrame frame)	Unknown
 	WindowsBase.dll!System.Windows.Threading.Dispatcher.PushFrame(System.Windows.Threading.DispatcherFrame frame)	Unknown
 	PresentationFramework.dll!System.Windows.Application.RunDispatcher(object ignore)	Unknown
 	PresentationFramework.dll!System.Windows.Application.RunInternal(System.Windows.Window window)	Unknown
 	PresentationFramework.dll!System.Windows.Application.Run(System.Windows.Window window)	Unknown
 	PresentationFramework.dll!System.Windows.Application.Run()	Unknown
 	WpfVisualizer.exe!WpfVisualizer.App.Main()	Unknown
```

### PR DESCRIPTION
**Pull request by jwoo-msft**
## Related Issue
Fixes #3976 

## Description
Initially, when template library is designed, the library supposed to return the payload unchanged if parsing fails. However, this is not possible if input payload is malformed. Especially when $data can not be parsed, recovery is not possible. Added error handling to take care of such cases. This change also added more detailed error messages, and throw error at the exact point of failure, so users can see user error immediately.

Also made changes to sample wpf app, so it can handle the exception from template library better by showing the error messages, and passing the input payload unchanged to AdaptiveCards parser, so it can try render the card.

## How Verified
How you verified the fix, including one or all of the following: 
1. All unit tests are passing
**Example of Error Message** when sample card from issue #3976 is used.
image

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/AdaptiveCards/pull/4011)

### CODE REVIEW

File: source/dotnet/Library/AdaptiveCards.Templating/AdaptiveCardsTemplateVisitor.cs [hunk 1]
```diff
@@ -92,8 +92,15 @@ public AdaptiveCardsTemplateVisitor(Func<string, object> nullSubstitutionOption,
...
-                root = JToken.Parse(data);
-                PushDataContext(data, root);
+                try
+                {
+                    root = JToken.Parse(data);
+                    PushDataContext(data, root);
+                }
+                catch (Exception e)
+                {
+                    throw new Exception("Setting root data failed with given data cotext", e);
```
Comment (paulcam206) [inline source/dotnet/Library/AdaptiveCards.Templating/AdaptiveCardsTemplateVisitor.cs:102]: nit:

```suggestion
                    throw new Exception("Setting root data failed with given data context", e);
```

File: source/dotnet/Library/AdaptiveCards.Templating/AdaptiveCardsTemplateVisitor.cs [hunk 2]
```diff
@@ -140,21 +147,18 @@ private void PushTemplatedDataContext(string jpath)
             DataContext parentDataContext = GetCurrentDataContext();
             if (jpath == null || parentDataContext == null)
             {
-                return;
+                throw new ArgumentNullException("parent data context or selection path is null");
```
Comment (paulcam206) [inline source/dotnet/Library/AdaptiveCards.Templating/AdaptiveCardsTemplateVisitor.cs:150]: nit:
```suggestion
                throw new ArgumentNullException("Parent data context or selection path is null");
```

**Comment by jwoo-msft** [comment]
@thanks Paul!

File: source/dotnet/Library/AdaptiveCards.Templating/AdaptiveCardsTemplateVisitor.cs (hunk 1 above)
Comment (jwoo-msft) [reply source/dotnet/Library/AdaptiveCards.Templating/AdaptiveCardsTemplateVisitor.cs:102]: :broken_heart: not again

---
In reply to: [424789404](https://github.com/microsoft/AdaptiveCards/pull/4011#discussion_r424789404) [](ancestors = 424789404)

**Pull request merged**